### PR TITLE
[Review]feat(plugin): add support for ECC_nistP256 security policy

### PR DESCRIPTION
**Pull request by zeschg**
Add http://opcfoundation.org/UA/SecurityPolicy#ECC_nistP256 policy support in OpenSSL plugin.

### CODE REVIEW

File: plugins/crypto/openssl/securitypolicy_common.c [hunk 1]
```diff
@@ -1307,18 +1318,19 @@ UA_OpenSSL_CreateSigningRequest(EVP_PKEY *localPrivateKey,
...
-    if (len > 1 && pkData[0] == 0x30 && pkData[1] == 0x82) { // Magic number for DER encoded keys
-        result = d2i_PrivateKey(EVP_PKEY_RSA, NULL,
-                                          &pkData, len);
-    } else {
-        BIO *bio = NULL;
+    /* Try to read DER encoded private key */
+    bio = BIO_new_mem_buf((void *) privateKey->data, (int) privateKey->length);
+    result = d2i_PrivateKey_bio(bio, NULL);
+    BIO_free(bio);
+
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_common.c:1331]: Reuse bio memory.

File: plugins/crypto/openssl/securitypolicy_common.c [hunk 2]
```diff
@@ -1419,3 +1432,652 @@ UA_OpenSSL_LoadLocalCertificate(const UA_ByteString *certificate, UA_ByteString
...
+    }
+
+    if(EVP_PKEY_derive(kctx, sharedSecretOut->data, &sharedSecretOut->length) <= 0) {
+        ret = UA_STATUSCODE_BADINTERNALERROR;
+        goto errout;
+    }
+
+errout:
+    UA_ByteString_clear(keyPublicRemoteEncoded);
+    UA_ByteString_delete(keyPublicRemoteEncoded);
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_common.c:1540]: _delete does _clear internally.

File: plugins/crypto/openssl/securitypolicy_common.c [hunk 3]
```diff
@@ -1419,3 +1432,652 @@ UA_OpenSSL_LoadLocalCertificate(const UA_ByteString *certificate, UA_ByteString
...
+UA_OpenSSL_ECDH (const int nid,
+                 EVP_PKEY * keyPairLocal,
+                 const UA_ByteString * keyPublicRemote,
+                 UA_ByteString * sharedSecretOut) {
+    UA_StatusCode ret = UA_STATUSCODE_GOOD;
+    EVP_PKEY_CTX * pctx = NULL;
+    EVP_PKEY_CTX * kctx = NULL;
+    EVP_PKEY * remotePubKey = NULL;
+    EVP_PKEY * params = NULL;
+    UA_ByteString * keyPublicRemoteEncoded = NULL;
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_common.c:1447]: This bytestring can be on the stack.
UA_ByteString mystr = UA_BYTESTRING_NULL;

File: plugins/crypto/openssl/securitypolicy_common.c [hunk 4]
```diff
@@ -1419,3 +1432,652 @@ UA_OpenSSL_LoadLocalCertificate(const UA_ByteString *certificate, UA_ByteString
...
+    if(keyPublicRemoteEncoded == NULL) {
+        ret = UA_STATUSCODE_BADOUTOFMEMORY;
+        goto errout;
+    }
+
+    if (UA_STATUSCODE_GOOD != UA_ByteString_allocBuffer(keyPublicRemoteEncoded, keyPublicRemote->length+1)) {
+        ret = UA_STATUSCODE_BADOUTOFMEMORY;
+        goto errout;
+    }
+
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_common.c:1459]: Add a comment about this magic number.

File: plugins/crypto/openssl/securitypolicy_common.c [hunk 5]
```diff
@@ -1419,3 +1432,652 @@ UA_OpenSSL_LoadLocalCertificate(const UA_ByteString *certificate, UA_ByteString
...
+}
+
+static UA_StatusCode
+UA_OpenSSL_ECC_GenerateSalt (const size_t L,
+                             const UA_ByteString * label,
+                             const UA_ByteString * key1,
+                             const UA_ByteString * key2,
+                             UA_ByteString * salt) {
+    size_t saltLen = sizeof(uint16_t) + label->length + key1->length + key2->length;
+
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_common.c:1556]: Salt == NULL is rather an internalerror than out of memory.

File: plugins/crypto/openssl/securitypolicy_common.c [hunk 6]
```diff
@@ -1419,3 +1432,652 @@ UA_OpenSSL_LoadLocalCertificate(const UA_ByteString *certificate, UA_ByteString
...
+    UA_Byte * ptmpData = signatureEnc->data;
+
+    signatureEnc->length = i2d_ECDSA_SIG(signatureRaw, &ptmpData);
+    if (signatureEnc->length == 0) {
+        ret = UA_STATUSCODE_BADINTERNALERROR;
+        goto errout;
+    }
+
+    mdctx = EVP_MD_CTX_create();
+    if(!mdctx) {
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_common.c:2018]: Goto errout

File: plugins/crypto/openssl/securitypolicy_eccnistp256.c [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+    /* delete all allocated members in the context */
+
+    Policy_Context_EccNistP256 *pc =
+        (Policy_Context_EccNistP256 *)policy->policyContext;
+    if (pc == NULL) {
+        return;
+    }
+
+    EVP_PKEY_free(pc->localPrivateKey);
+    pc->localPrivateKey = NULL;
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_eccnistp256.c:103]: No need to zero out if you free after.

File: plugins/crypto/openssl/securitypolicy_eccnistp256.c [hunk 8]
```diff
@@ -0,0 +1,722 @@
...
+
+/* create the channel context */
+
+static UA_StatusCode
+UA_ChannelModule_EccNistP256_New_Context(const UA_SecurityPolicy *securityPolicy,
+                                                 const UA_ByteString *remoteCertificate,
+                                                 void **channelContext) {
+    if(securityPolicy == NULL || remoteCertificate == NULL || channelContext == NULL) {
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
```
Comment (jpfr) [inline plugins/crypto/openssl/securitypolicy_eccnistp256.c:192]: Use UA_calloc.
Then the memory is already initialized to zero.
And you can remove the _init.

**Review by jpfr — COMMENTED** [review]
Nice!
Minor comments inline.
Can be merged then.

File: plugins/crypto/openssl/securitypolicy_common.c (hunk 1 above)
Comment (zeschg) [reply plugins/crypto/openssl/securitypolicy_common.c:1331]: Already done in #6809

**Comment by jpfr** [comment]
Nice! Merged.

**Pull request merged**